### ISSUE TITLE
[Feature Request] Add clipboard support

### ISSUE THREAD

**Issue by gezihuzi** [proxin187/Yaxi#1]
Hello! I'm interested in using Yaxi for my project and wondering about clipboard functionality.

Does Yaxi have plans to support X11 clipboard operations (like CLIPBOARD and PRIMARY selection)? This would include basic operations like:
- Copy text/data to clipboard
- Paste from clipboard
- Monitor clipboard changes
- Handle different selection types

If there's no immediate plan for this feature, I'd be willing to contribute to implementing it. Would you be open to accepting a PR for clipboard support?

Thanks for creating this promising X11 Rust library!

**Comment by proxin187** [comment]
Thanks for the issue!

Clipboard support is definitely a feature needed, it shouldn't be a problem for me to implement it so ill try to get it done today!

Also, error reporting is a little wacky right now so ill probably make it better too.

**Comment by gezihuzi** [comment]
Thanks for the quick response! That's great to hear.

I'd be happy to help with testing and provide feedback once the implementation is ready. I can try different use cases and report any issues or edge cases I find.

Please let me know when there's something ready for testing.

**Comment by proxin187** [comment]
Hello, the clipboard implementation should be working now.

Its gated behind the "clipboard" feature and there should be an example of how to use it inside the examples folder.

Currently it only supports UTF8_STRING but adding support for other types shouldn't be hard!

Please let me know if there is anything missing!

**Comment by gezihuzi** [comment]
Thanks for the implementation! I've noticed some unexpected behavior while testing the clipboard feature:

Observed behavior:
1. When I copy text from other applications -> run paste example -> successfully read the content
2. When I run copy example (writes predefined text) -> shows success -> run paste example -> fails to read the content

I've also added three unit tests in PR #2  to verify basic clipboard operations:
- Read text from clipboard
- Write text to clipboard
- Write-then-read consistency check

However, these tests are not passing as expected. 

Could you help take a look at what might be causing these issues? I can provide more details or add debug logging if needed.

**Comment by proxin187** [comment]
Thanks so much for the feedback and PR.

The issue where you couldn't read the clipboard after running copy is most likely because you lose x11 selections after the selection owner exits, on my machine it works as long as the copy example is still running while the paste example is run.

[Short summary of x11 selection](https://www.uninformativ.de/blog/postings/2017-04-02/0/POSTING-en.html)

Example here:
![example gif](https://github.com/user-attachments/assets/fe1b9b89-b9b5-4a0c-9652-c9f9472185eb)

When it comes to your tests you made, ill try to debug them and get them working as soon as possible, (im at school right now so it might not be done before later today)

Thanks for letting me know!

**Comment by proxin187** [comment]
Hello, Sorry for the inconveniences but the clipboard is finally in a semi-stable state, i had to refactor the concurrency model in yaxi so it took a little longer then i would have liked but it is done.

feel free to submit  PR's for more clipboard formats if you need them.

**Comment by gezihuzi** [comment]
Thank you for improving the clipboard, The implementation is working much better now.

I plan to submit new PRs to enhance the clipboard support, including:

- [ ] Support for additional clipboard formats (imges, files, custom mime)
- [ ] Monitor clipboard changes
- [ ] More comprehensive test coverage
- [ ] Potential improvements to clipboard operations

I'll create separate PRs for each enhancement to keep changes focused and manageable. Looking forward to contributing more to this project.

**Comment by proxin187** [comment]
Thanks for your great PR, these features where much needed.

## Answers

1. Could you please make an example that reproduces this behavior? i wasnt able to reproduce it myself, but if i get a example i will most likely be able to fix it quite fast.
2. If you meant if we should clean/empty the ClipboardData structure once we lose the selection ownership, then yes but it isn't extremely important. On the other hand if you meant when our process exits and the clipboard structure is dropped then i think we should destroy the window and delete the atoms that are only used by us eg. (storage property).
3. When it comes to the clipboard manager i havent done too much research yet, but when it comes to saving properly when the process terminates, you could most likely just fit some sort of a save mechanism inside the Drop implementation for Clipboard like this:
```rust
impl Drop for Clipboard {
    fn drop(&mut self) {
        // SOME SORT OF A SAVE MECHANISM HERE

        self.listener.kill();
    }
}
```

Ill most likely have to look a little more into the clipboard manager though!

4. There is definitely big room for improvement when it comes to the current concurrency model of our clipboard implementation, and im a 100% in favor of exploring different architectures and trying to find the best one.

## Unrelated Sidenote

Also just as a kind of unrelated side note, the queue implementation has gotten a massive performance boost as i switched it to using [Condvar](https://doc.rust-lang.org/std/sync/struct.Condvar.html) when waiting for events instead of the old method which would simply busy loop until the queue wasnt empty.

**Comment by gezihuzi** [comment]
1. Debugging at a specific line will result in an error.
![image](https://github.com/user-attachments/assets/e19219e2-a4e2-46e6-9438-7d5baff5a9f7)
2 & 3. I think we can store the content in `CLIPBOARD_MANAGER` in the `Drop` implementation and then clean up the data.

**Comment by proxin187** [comment]
I've implemented saving to the clipboard manager once the clipboard structure is dropped and also cleaning up the selection structure when we lose the selection,  there are a few notable things with my implementation:

1. Once the `Clipboard` structure is dropped it will ask the clipboard manager to save content using the `SAVE_TARGETS` atom then it will wait for a request from the clipboard manager with a timeout of 100ms, the potential problem with this is that this will result in a 100ms wait on all systems that dont have a clipboard manager running, im not sure if this is a problem that can be solved as when i looked into other clipboard implementations they seem to do the same thing notably [arboard's implementation](https://github.com/1Password/arboard/blob/master/src/platform/linux/x11.rs#L707).

i would like to hear your opinion on this and if you have any proposals for a better way to do this!

**Comment by gezihuzi** [comment]
@proxin187 

Thank you for completing the `CLIPBOARD_MANAGER` implementation. I think waiting for 100ms at this stage is a feasible solution. In the future, maybe we can check if the current `CLIPBOARD_MANAGER` exists in advance so that we can skip it directly, which may be more efficient?

Share with you some of my recent progress. I am currently working on refactoring the implementation of the `clipboard` module in my [temporary work branch](https://github.com/gezihuzi/Yaxi/tree/feature/clipboard-manager). But it's not finished yet.

The purpose of refactoring is not only to enhance support for different `selection` (PRIMARY, SECONDARY, CLIPBOARD), but also to support the reading and writing of different types of clipboard data; advanced features (CLIPBOARD_MANAGER and INCR etc.); optimize event polling and shared data reading and writing processing.

Please feel free to share your thoughts and suggestions on the part of my work that I am currently working on.

**Comment by gezihuzi** [comment]
@proxin187 

Hi, I've now completed the clipboard refactoring we discussed earlier. This [PR](#8 ) implements the enhanced clipboard functionality with support for different X11 selections and data formats. Would love to get your thoughts on this implementation. Thank you!

**Comment by proxin187** [comment]
@gezihuzi

Hi, thank you so much for the PR, i had to fix some issues i had with the tests but got it to work after a while, but i do have some problems to discuss, notably the tests seem to use a unreasonable long time to complete, im not sure why this is but it would be great if you could figure it out, other than that i find your clipboard implementation to be very good and i am looking forward to expanding it!

**Comment by gezihuzi** [comment]
> @gezihuzi
> 
> Hi, thank you so much for the PR, i had to fix some issues i had with the tests but got it to work after a while, but i do have some problems to discuss, notably the tests seem to use a unreasonable long time to complete, im not sure why this is but it would be great if you could figure it out, other than that i find your clipboard implementation to be very good and i am looking forward to expanding it!

Thank you for the feedback! I'm glad you find the clipboard implementation useful.

Regarding the long test execution time - I'll definitely look into this. Could you share:
1. Which specific tests are taking longer than expected?
2. What kind of execution time are you seeing?

This will help me investigate and optimize the test performance. I'm also happy to collaborate on expanding the clipboard functionality further!

**Comment by proxin187** [comment]
@gezihuzi

thanks for the quick response, after further investigation i found that its the `test_clipboard_clear` test that fails, the issue seems to be that the first `clipboard.get_targets` call results in a timeout.

**Comment by gezihuzi** [comment]
@proxin187 

The current implementation of clear has known defects and may not achieve the expected results after execution. I think replacing the internal implementation with 'setting the content to an empty UTF-8 string' might achieve a similar effect.

Also, I've left some comments about test cases in the PR discussion thread - would appreciate your thoughts on those when you have a chance.

https://github.com/proxin187/Yaxi/commit/e603bd51e70a94be2256cb4b771d1bbb9c3c315c#commitcomment-149780469

**Comment by gezihuzi** [comment]
@proxin187 

Hi, I've submitted a new PR #9  to fix the `clear` operation failing to empty the clipboard. I've also added support for reading and writing `uri_list` and `image` clipboard types.

Additionally, I've improved the test cases based on our previous discussion at https://github.com/proxin187/Yaxi/commit/e603bd51e70a94be2256cb4b771d1bbb9c3c315c#commitcomment-149818973

Could you please review the changes and let me know if you have any suggestions for improvement?
